tooi/api/streaming.py

```diff
@@ -0,0 +1,392 @@
+"""
+Streaming (push) support for timelines.
+
+Streaming is handled by the InstanceStreamer class, which should only be created once per instance.
+InstanceStreamer handles connecting to the streaming endpoint, receiving events, and distributing
+them to the various subscribers.  To subscribe to streaming, use subscribe() to create a
+StreamSubscription instance.
+
+https://docs.joinmastodon.org/methods/streaming/
+"""
+
+import aiohttp
+import asyncio
+import json
+import logging
+import re
+import urllib
+
+from tooi.asyncio import run_async_task
+from tooi.context import get_context
+from tooi.data.instance import InstanceInfo
+
+
+class StreamEvent:
+    """Internal type representing an event received from a stream."""
+    def __init__(self, stream: str, event: str, payload):
+        self.stream = stream
+        self.event = event
+        self.payload = payload
+
+
+StreamQueue = asyncio.Queue[StreamEvent]
+logger = logging.getLogger(__name__)
+
+
+class WSStreamClient:
+    """
+    Client for connecting to a streaming endpoint and fetching events.
+    This implements the WebSocket protocol.
+    """
+
+    TIMEOUT = 120
+    ERROR_BACKOFF = 30
+
+    def __init__(
+            self,
+            instance: InstanceInfo,
+            url: str,
+            stream_name: str):
+
+        # TODO: This should be part of the instance.
+        self.ctx = get_context()
+        self.instance = instance
+        self.url = url
+        self.stream_name = stream_name
+        self.queue = StreamQueue()
+        self._lines = []
+
+    async def close(self):
+        pass
+
+    async def run(self):
+        logger.info(f"WSStreamClient: running url={self.url}")
+
+        # Run forever, silently reconnecting if we get an error.
+        while True:
+            try:
+                await self._stream()
+            except aiohttp.ClientResponseError as exc:
+                logger.info((
+                    f"WSStreamClient: disconnected from stream={self.stream_name}: "
+                    f"{exc.status} {exc.message}"))
+            except aiohttp.ClientError as exc:
+                logger.info((
+                    f"WSStreamClient: disconnected from stream={self.stream_name}: "
+                    f"{exc}"))
+
+            await asyncio.sleep(self.ERROR_BACKOFF)
+
+    async def _stream(self):
+        logger.info(f"WSStreamClient: connecting to stream={self.stream_name}")
+
+        timeout = aiohttp.ClientTimeout(
+                total=None,
+                connect=10,
+                sock_connect=10,
+                sock_read=self.TIMEOUT
+        )
+
+        # Since the base URL of the stream might be different from the instance's normal base URL,
+        # we have to create our own client.
+        client = aiohttp.ClientSession()
```

Comment:
Why not use the context manager here as they recommend?

```py
async with aiohttp.ClientSession() as session:
```

Reply:
because i forgot :-)  fixed in 5b2b6ca